pkg/applicationserver/io/pubsub/pubsub.go

```diff
@@ -17,119 +17,280 @@ package pubsub
+		case errors.IsCanceled(err),
+			errors.IsAlreadyExists(err):
+			return nil
+		default:
+			return err
 		}
-		sub := io.NewSubscription(s.ctx, "pubsub", nil)
-		// Publish upstream
-		go func() {
-			logger := log.FromContext(s.ctx).WithField("publish-url", pubURL)
-			for {
-				select {
-				case <-sub.Context().Done():
-					logger.WithError(sub.Context().Err()).Debug("Done sending upstream messages")
-					return
-				case up := <-sub.Up():
-					buf, err := s.formatter.FromUp(up.ApplicationUp)
-					if err != nil {
-						log.WithError(err).Warn("Failed to marshal upstream message")
-						continue
-					}
-					err = topic.Send(ctx, &pubsub.Message{
-						Body: buf,
-					})
-					if err != nil {
-						log.WithError(err).Warn("Failed to publish upstream message")
-						continue
-					}
-					logger.Debug("Publish upstream message")
-				}
+	}, component.TaskRestartOnFailure, 0.1, integrationBackoff...)
+}
+
+type integration struct {
+	ttnpb.ApplicationPubSub
+	ctx    context.Context
+	cancel errorcontext.CancelFunc
+
+	conn *provider.Connection
+
+	server io.Server
+	sub    *io.Subscription
+	format Format
+}
+
+func (i *integration) handleUp(ctx context.Context) {
+	logger := log.FromContext(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			logger.WithError(ctx.Err()).Debug("Done sending upstream messages")
+			return
+		case up := <-i.sub.Up():
+			var topic *pubsub.Topic
+			switch up.ApplicationUp.Up.(type) {
+			case *ttnpb.ApplicationUp_UplinkMessage:
+				topic = i.conn.Topics.UplinkMessage
+			case *ttnpb.ApplicationUp_JoinAccept:
+				topic = i.conn.Topics.JoinAccept
+			case *ttnpb.ApplicationUp_DownlinkAck:
+				topic = i.conn.Topics.DownlinkAck
+			case *ttnpb.ApplicationUp_DownlinkNack:
+				topic = i.conn.Topics.DownlinkNack
+			case *ttnpb.ApplicationUp_DownlinkSent:
+				topic = i.conn.Topics.DownlinkSent
+			case *ttnpb.ApplicationUp_DownlinkFailed:
+				topic = i.conn.Topics.DownlinkFailed
+			case *ttnpb.ApplicationUp_DownlinkQueued:
+				topic = i.conn.Topics.DownlinkQueued
+			case *ttnpb.ApplicationUp_LocationSolved:
+				topic = i.conn.Topics.LocationSolved
+			}
+			if topic == nil {
+				continue
+			}
+			buf, err := i.format.FromUp(up.ApplicationUp)
+			if err != nil {
+				logger.WithError(err).Warn("Failed to marshal upstream message")
+				continue
 			}
-		}()
-		subs = append(subs, sub)
-		s.topics = append(s.topics, topic)
+			err = topic.Send(ctx, &pubsub.Message{
+				Body: buf,
+			})
+			if err != nil {
+				logger.WithError(err).Warn("Failed to publish upstream message")
+				continue
+			}
+			logger.Debug("Publish upstream message")
+		}
 	}
-	for _, subURL := range subURLs {
-		subscription, err := pubsub.OpenSubscription(ctx, subURL)
+}
+
+func (i *integration) handleDown(ctx context.Context, op func(io.Server, context.Context, ttnpb.EndDeviceIdentifiers, []*ttnpb.ApplicationDownlink) error, subscription *pubsub.Subscription) {
+	logger := log.FromContext(ctx)
+	for ctx.Err() == nil {
```

Comment:
Early exit

Reply:
I.e. break on error ? Since we cannot do `select` on `Receive` unfortunately.

Reply:
Sorry bad comment of mine